File: windows/AudioTranscriptionSummary/Services/WaveformDataProvider.cs
#nullable enable
using System;
using NAudio.Wave;

namespace AudioTranscriptionSummary.Services;

/// <summary>
/// 波形描画用のサンプルデータを提供する
/// 音声ファイルからサンプルを読み込み、ダウンサンプリングして正規化した配列を返す
/// </summary>
public static class WaveformDataProvider
{
    /// <summary>
    /// 音声ファイルから波形データを抽出する
    /// </summary>
    /// <param name="filePath">音声ファイルのパス</param>
    /// <param name="sampleCount">出力するサンプル数（描画解像度、デフォルト: 200）</param>
    /// <returns>正規化された振幅値の配列（0.0〜1.0）。エラー時は空配列を返す</returns>
    public static float[] LoadWaveformData(string filePath, int sampleCount = 200)
    {
        if (sampleCount <= 0) return Array.Empty<float>();

        try
        {
            using var reader = new AudioFileReader(filePath);
            var totalSamples = (int)(reader.Length / (reader.WaveFormat.BitsPerSample / 8));
            if (totalSamples <= 0) return Array.Empty<float>();

            // 全サンプルを読み込む
            var buffer = new float[totalSamples];
            int samplesRead = reader.Read(buffer, 0, totalSamples);
            if (samplesRead <= 0) return Array.Empty<float>();

            int channels = reader.WaveFormat.Channels;
            int frameCount = samplesRead / channels;
            if (frameCount <= 0) return Array.Empty<float>();

            // 全フレームを sampleCount 個のビンに分割し、各ビンの最大振幅を取得
            var amplitudes = new float[sampleCount];

            for (int bin = 0; bin < sampleCount; bin++)
            {
                int start = bin * frameCount / sampleCount;
                int end = Math.Min((bin + 1) * frameCount / sampleCount, frameCount);
                if (start >= end) continue;

                float maxAmplitude = 0;
                for (int i = start; i < end; i++)
                {
                    // 全チャンネルの最大振幅を取得
                    for (int ch = 0; ch < channels; ch++)
                    {
                        float sample = Math.Abs(buffer[i * channels + ch]);
                        if (sample > maxAmplitude)
                            maxAmplitude = sample;
                    }
                }
                amplitudes[bin] = maxAmplitude;
            }

            // 全体の最大値で正規化（0.0〜1.0）
            float globalMax = 0;
            for (int i = 0; i < amplitudes.Length; i++)
            {
                if (amplitudes[i] > globalMax)
                    globalMax = amplitudes[i];
            }

            if (globalMax > 0)
            {
                for (int i = 0; i < amplitudes.Length; i++)
                {
                    amplitudes[i] /= globalMax;
                }
            }

            return amplitudes;
        }
        catch
        {
            // エラー時は空配列を返す
            return Array.Empty<float>();
        }
    }
}

File: macos/Sources/AudioTranscriptionSummary/Services/WaveformDataProvider.swift
// WaveformDataProvider.swift
// 音声ファイルから波形描画用のサンプルデータを抽出するプロバイダー
// AVAudioFile でサンプルを読み込み、ダウンサンプリングして正規化した配列を返す

import AVFoundation

/// 波形描画用のサンプルデータを提供する
struct WaveformDataProvider {

    /// 音声ファイルから波形データを抽出する
    /// - Parameters:
    ///   - url: 音声ファイルの URL
    ///   - sampleCount: 出力するサンプル数（描画解像度、デフォルト: 200）
    /// - Returns: 正規化された振幅値の配列（0.0〜1.0）。エラー時は空配列を返す
    static func loadWaveformData(from url: URL, sampleCount: Int = 200) -> [Float] {
        guard sampleCount > 0 else { return [] }

        // AVAudioFile でファイルを開く
        let audioFile: AVAudioFile
        do {
            audioFile = try AVAudioFile(forReading: url)
        } catch {
            return []
        }

        let totalFrames = AVAudioFrameCount(audioFile.length)
        guard totalFrames > 0 else { return [] }

        // PCM バッファにサンプルを読み込む
        guard let format = AVAudioFormat(commonFormat: .pcmFormatFloat32,
                                         sampleRate: audioFile.processingFormat.sampleRate,
                                         channels: audioFile.processingFormat.channelCount,
                                         interleaved: false),
              let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: totalFrames) else {
            return []
        }

        do {
            try audioFile.read(into: buffer)
        } catch {
            return []
        }

        guard let channelData = buffer.floatChannelData else { return [] }
        let frameCount = Int(buffer.frameLength)
        guard frameCount > 0 else { return [] }

        // 全サンプルを sampleCount 個のビンに分割し、各ビンの最大振幅を取得
        let samplesPerBin = max(frameCount / sampleCount, 1)
        var amplitudes = [Float](repeating: 0, count: sampleCount)

        for bin in 0..<sampleCount {
            let start = bin * frameCount / sampleCount
            let end = min((bin + 1) * frameCount / sampleCount, frameCount)
            guard start < end else { continue }

            var maxAmplitude: Float = 0
            for i in start..<end {
                // 全チャンネルの最大振幅を取得
                for ch in 0..<Int(format.channelCount) {
                    let sample = abs(channelData[ch][i])
                    if sample > maxAmplitude {
                        maxAmplitude = sample
                    }
                }
            }
            amplitudes[bin] = maxAmplitude
        }

        // 全体の最大値で正規化（0.0〜1.0）
        let globalMax = amplitudes.max() ?? 0
        guard globalMax > 0 else { return amplitudes }

        for i in 0..<amplitudes.count {
            amplitudes[i] /= globalMax
        }

        return amplitudes
    }
}
